src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcServerTransport.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    internal abstract class IpcServerTransport : IDisposable
+    {
+        public static IpcServerTransport Create(string transportPath)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return new WindowsPipeServerTransport(transportPath);
+            }
+            else
+            {
+                return new UnixDomainSocketServerTransport(transportPath);
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+        }
+
+        public abstract Task<Stream> AcceptAsync(CancellationToken token);
+    }
+
+    internal sealed class WindowsPipeServerTransport : IpcServerTransport
+    {
+        private const string PipePrefix = @"\\.\pipe\";
+
+        private bool _disposed = false;
+        private NamedPipeServerStream _stream;
+
+        private readonly string _pipeName;
+
+        public WindowsPipeServerTransport(string pipeName)
+        {
+            _pipeName = pipeName.StartsWith(PipePrefix) ? pipeName.Substring(PipePrefix.Length) : pipeName;
+            CreateNewPipeServer();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                    _stream.Dispose();
+                }
+                _disposed = true;
+            }
+        }
+
+        public override async Task<Stream> AcceptAsync(CancellationToken token)
+        {
+            NamedPipeServerStream connectedStream;
+            try
+            {
+                await _stream.WaitForConnectionAsync(token);
+
+                connectedStream = _stream;
+            }
+            finally
+            {
+                CreateNewPipeServer();
+            }
+            return connectedStream;
+        }
+
+        private void CreateNewPipeServer()
+        {
+            _stream = new NamedPipeServerStream(_pipeName, PipeDirection.InOut, 10, PipeTransmissionMode.Byte, PipeOptions.Asynchronous);
```

Comment:
I'm wondering if this instance limit is too low.

Reply:
Yes, definitely too low. I don't know if I want to allow it to be infinite, but it needs to be significantly higher. A single runtime instance can have many connections to the same reversed server (1 for each active diagnostics session + 1 more for new connection).